### ISSUE TITLE
Occasional unstable behaviour with company and completions

### ISSUE THREAD

**Issue by Zulu-Inuoe** [joaotavora/sly#210]
Likely related to #209 

Test case:

Start sly and have two windows in the same frame active. Now begin a completion such as `(ql:` and execute `company-select-next` (I have it bound to `C-n` and hold it down).
At random times, the company selection will be cancelled, and the cursor will jump to the other window.

If I've two mrepl windows active, it'll endlessly jump from one to another and attempt to continue completion there, endlessly cycling between the two.

At best I've also found it just randomly issuing a `Quit` event somewhere and booting me out of the `company-active` mode

I found that removing all the company callback functions from `sly--completions-complete-symbol-1` prevents this issue altogether, and if I uncomment only `:company-docsig`, the behaviour reappears

Software:
Emacs 26.1, Windows 10 x64, SBCL 1.4.12

**Comment by Zulu-Inuoe** [comment]
From what I've found, the cursor jumping windows only occurs when `company-quickhelp` is also installed and enabled.
However, getting booted out of `company` occurs even on an Emacs with just `company` and `sly` installed

**Comment by joaotavora** [comment]
I'm going to close this in favor of #209, since I think it's basically the same problem.

**Comment by Zulu-Inuoe** [comment]
Hi.

Using the given recipe, I was able to reproduce it 'Randomly Cancelling'.

By 'random' I do mean unexpected, yes. That is, if suggestions list from company is active, holding down `<down>` (default binding) will cause company, at a 'random amount of time later' to be disable,
I'm testing this by typing `(ql:` then, once company shows, pressing and holding `<down>` to go through the results. Note that by default, `<down>` is bound to `company-select-next-or-abort`, which will abort at the bottom of the list, but on `(ql:` I usually don't even reach the halfway point.

I've also tried binding to `company-select-next` (which does not abort) with the same results

To get the random aborts, I get them regardless of how many windows I have active. Such as having a single Emacs root window with `*sly-mrepl` buffer on it, typing `(ql:`, waiting for results list to show, and holding down `<down>`

I've recorded a clip of that happening. Here I started emacs as per your recipe, got to the sly mrepl, and recorded. I'm only holding `<down>` after typing:

![gif](https://user-images.githubusercontent.com/1526310/50559907-4e1d1780-0cc9-11e9-8c7f-7c3c9f65e0ce.gif)

if `company-quickhelp-delay` is set to 0 this happens practically instantaneously (within the first two or three symbols)

For it jumping windows, I was unable to replicate this this recipe, so I'll need to dig further to figure out what was causing that.

**Comment by joaotavora** [comment]
Hi @Zulu-Inuoe , thanks for confirming this.

I was able to confirm it too.  Unfortunately, it only happens in an optimized build of Emacs, so homing in on the (presumably Emacs's) bug, is going to be hard.  I'll try a bit harder before I report this to `bug-gnu-emacs@gnu.org`.

**Comment by Zulu-Inuoe** [comment]
Yes at this point I am pretty convinced it's not a problem with sly, but with `company-capf`, `company`, or maybe even Emacs itself.
I wonder if other company backends with 'long' delays retrieving information (async process) have similar issues.

And sorry, I should have been more clear: I was unable to reproduce jumping windows even with the previous commit (61e45f75b0996ad47052ccca41f84d031ddad515) using this recipe, so it's likely something else with my usual emacs setup that a factor in getting it to happen

**Comment by joaotavora** [comment]
I've only now noticed your `company-quickhelp-delay` thing.  I'm going to try that and reproducing again...

**Comment by joaotavora** [comment]
OK, I have something. When the bug happens in the very latest SLY commit, it's not exactly an  unexpected `quit` signal, seems an interactive call to `company-abort`. But I don't know where it's coming from, since I didn't press C-g.

```
Debugger entered--entering a function:
* company-cancel(abort)
  company-abort()
  funcall-interactively(company-abort)
  call-interactively(company-abort nil nil)
  command-execute(company-abort)
```

**Comment by joaotavora** [comment]
However, in that Lisp stack frame it does say that `last-input-event` is `7`, which is `C-g`.

**Comment by Zulu-Inuoe** [comment]
You've had better luck with it than I then, because the most I've been able to see (after doing `M-x toggle-debug-on-quit`) is:
![image](https://user-images.githubusercontent.com/1526310/50574354-f8259e00-0db4-11e9-8cfb-e499ae71680f.png)

**Comment by joaotavora** [comment]
I'm not using `toggle-debug-on-quit`, I'm using `M-x debug-on-entry RET company-cancel RET`.

And I'm putting a GDB breakpoint on `quit()` and still no quit.  If I use `toggle-debug-on-quit` I don't see a quit at all.

**Comment by joaotavora** [comment]
There may be some differences between W32 and linux here.  I don't have time to dig much deeper now, so I'm going to mark this "emacs bug" and "low priority"  Let me know if this bothers you a lot in real life (do you really set `company-quickhelp-delay` to 0?)

**Comment by Zulu-Inuoe** [comment]
Understood. This all seems like a very complicated bug and maybe not a priority.

Unfortunately, it occurs even if I completely disable `company-quickhelp`, so it does bother me quite a bit. What I'm doing right now is manually commenting out all the company integration in `lib/sly-completions.el`.

That way I can at least have the list of symbols (no doc or match %).

**Comment by joaotavora** [comment]
> Unfortunately, it occurs even if I completely disable company-quickhelp, so it does bother me quite a bit. What I'm doing right now is manually commenting out all the company integration in lib/sly-completions.el.

It would probably be easier to disable company in `sly-mode-hook`.

I'm trying on Windows now and I still can't reproduce the problem.  Can you try with a more recent Emacs?

**Comment by Zulu-Inuoe** [comment]
Well, I still make use of company for inline completion, and by disabling the lookup of doc info etc I still get the company list of available completions, just no doc or other metadata.

What version of Emacs are you using? I'm making use of 26.1 x64 from one of the [GNU Mirrors](https://mirror.leifrogers.com/gnu/emacs/windows/emacs-26/).

I -did- try '[emax](https://github.com/m-parashar/emax64)' briefly yesterday and it -seemed- to have fixed the problem, but I was running into other issues so I couldn't test further. I'll keep you updated on that.

Thank you.

**Comment by joaotavora** [comment]
Hi @Zulu-Inuoe, just want to report that after https://debbugs.gnu.org/cgi/bugreport.cgi?bug=34394, the problem is most likely fixed.  The issues you were seeing were related to a deep-seated Emacs bug when company was used. That bug has been fixed and the fix is available in the Git master and in the upcoming 26.2 version, in case you want to try it out. 

I'm going to tentatively close this.

**Comment by Zulu-Inuoe** [comment]
Thank you very much for the update! I never got around to trying the other distro, but it sounds like that shouldn't have made much difference.
